{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "27444ced",
   "metadata": {},
   "source": [
    "# AMPSphere v.2022-03\n",
    "\n",
    "This is a notebook meant to form the set of notebooks used to analyze the data in AMPSphere and write the manuscript:\n",
    "\n",
    "__AMPSphere: Global survey of prokaryotic antimicrobial peptides shaping microbiomes__\n",
    "\n",
    "### Summarizing AMPSphere origins and results in supplementary tables\n",
    "\n",
    "Here, we summarize the metagenomes, genomes and other info needed to rebuild AMPSphere along with other supplementary data reporting results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "061c4188",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading libraries\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import spearmanr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5f8516a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "higher_level = {'sediment' : 'other',\n",
    "        'bird gut' : 'other animal',\n",
    "        'cat gut' : 'non-human mammal gut',\n",
    "        'insect associated' : 'other animal',\n",
    "        'human urogenital tract' : 'other human',\n",
    "        'dog gut' : 'non-human mammal gut',\n",
    "        'fermented food' : 'anthropogenic',\n",
    "        'groundwater' : 'aquatic',\n",
    "        'coral associated' : 'other animal',\n",
    "        'rat gut' : 'non-human mammal gut',\n",
    "        'human associated' : 'other human',\n",
    "        'cattle gut' : 'non-human mammal gut',\n",
    "        'deer gut' : 'non-human mammal gut',\n",
    "        'mouse gut' : 'non-human mammal gut',\n",
    "        'river associated' : 'aquatic',\n",
    "        'primate gut' : 'non-human mammal gut',\n",
    "        'human respiratory tract' : 'other human',\n",
    "        'cattle rumen' : 'other animal',\n",
    "        'human saliva' : 'other human',\n",
    "        'activated sludge' : 'anthropogenic',\n",
    "        'lake associated' : 'aquatic',\n",
    "        'wastewater' : 'anthropogenic',\n",
    "        'chicken gut' : 'other animal',\n",
    "        'air' : 'other',\n",
    "        'human mouth' : 'other human',\n",
    "        'plant associated' : 'soil/plant',\n",
    "        'water associated' : 'aquatic',\n",
    "        'pig gut' : 'non-human mammal gut',\n",
    "        'human skin' : 'other human',\n",
    "        'marine' : 'aquatic',\n",
    "        'soil' : 'soil/plant',\n",
    "        'built environment' : 'anthropogenic',\n",
    "        'human gut' : 'human gut',\n",
    "        'anthropogenic': 'anthropogenic',\n",
    "        'bear gut' : 'non-human mammal gut',\n",
    "        'bee gut': 'other animal',\n",
    "        'bat gut': 'non-human mammal gut',\n",
    "        'dog associated': 'other animal',\n",
    "        'cattle associated': 'other animal',\n",
    "        'crustacean associated': 'other animal',\n",
    "        'insect gut': 'other animal',\n",
    "        'goat gut': 'non-human mammal gut', \n",
    "        'rodent gut': 'non-human mammal gut',\n",
    "        'fisher gut': 'non-human mammal gut',\n",
    "        'human digestive tract': 'other human',\n",
    "        'coyote gut': 'non-human mammal gut',\n",
    "        'planarian associated': 'other animal',\n",
    "        'sponge associated': 'other animal',\n",
    "        'goat rumen': 'other animal',\n",
    "        'crustacean gut': 'other animal',\n",
    "        'annelidae associated': 'other animal',\n",
    "        'bird skin': 'other animal',\n",
    "        'beatle gut': 'other animal',\n",
    "        'termite gut': 'other animal', \n",
    "        'fish gut': 'other animal',\n",
    "        'mollusc associated': 'other animal',\n",
    "        'ship worm associated': 'other animal',\n",
    "        'rabbit gut': 'non-human mammal gut',\n",
    "        'tunicate associated': 'other animal',\n",
    "        'mussel associated': 'other animal',\n",
    "        'horse gut': 'non-human mammal gut',\n",
    "        'wasp gut': 'other animal',\n",
    "        'guinea pig gut': 'non-human mammal gut'}\n",
    "\n",
    "\n",
    "is_host_associated = {'human gut' : True,\n",
    "        'soil/plant' : False,\n",
    "        'aquatic' : False,\n",
    "        'anthropogenic' : False,\n",
    "        'other human' : True,\n",
    "        'non-human mammal gut' : True,\n",
    "        'other animal' : True,\n",
    "        'other' : False}\n",
    "\n",
    "\n",
    "plants={'lettuce', 'monocots', 'cowpea',\n",
    "        'mosses', 'pitcher plant', 'maize',\n",
    "        'thale cress', 'siratro', 'grapevine',\n",
    "        'Norway spruce', 'black cottonwood',\n",
    "        'soy', 'french bean', 'silvergrass',\n",
    "        'sorghum', 'bread wheat', 'sunflower',\n",
    "        'carrot', 'lodgepole pine' 'burclover',\n",
    "        'cottongrass', 'switchgrass', 'eudicots',\n",
    "        'agave', 'barrelclover', 'alfalfa',\n",
    "        'red fir'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "920c8172",
   "metadata": {},
   "outputs": [],
   "source": [
    "def listing(x):\n",
    "    return ', '.join(x.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4962502d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data = pd.read_table('data/gmsc_amp_genes_envohr_source.tsv.gz')\n",
    "\n",
    "data = data[data.is_metagenomic == True]\n",
    "\n",
    "data['higher'] = data['general_envo_name'].map(lambda g: higher_level.get(g, 'other'))\n",
    "\n",
    "data['host_associated'] = data['higher'].map(lambda g: is_host_associated.get(g, 'NA'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d6c038d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "host_associated\n",
      "False    595345\n",
      "True     285393\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "nf = data[['amp','host_associated']]\n",
    "nf = nf.drop_duplicates()\n",
    "nf = nf.groupby('host_associated')\n",
    "nf = nf.agg('size')\n",
    "print(nf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "844813f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "higher\n",
      "aquatic       213861\n",
      "soil/plant    315628\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "nf = data[data.higher.isin(['soil/plant', 'aquatic'])]\n",
    "nf = nf[['amp', 'higher']]\n",
    "nf = nf.drop_duplicates()\n",
    "nf = nf.groupby('higher')\n",
    "nf = nf.agg('size')\n",
    "print(nf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "96225ef0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "higher\n",
      "anthropogenic    85988\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "nf = data[data.higher == 'anthropogenic']\n",
    "nf = nf[['amp','higher']]\n",
    "nf = nf.drop_duplicates()\n",
    "nf = nf.groupby('higher')\n",
    "nf = nf.agg('size')\n",
    "print(nf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f03eee86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Other environments: 15609\n"
     ]
    }
   ],
   "source": [
    "nf = len(set(data[(data.higher == 'other')]['amp']))\n",
    "print(f'Other environments: {nf}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f6522c1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "host_common_name\n",
       "guinea pig                1\n",
       "rabbit                    1\n",
       "white-tailed deer         2\n",
       "cottongrass               2\n",
       "soy                       4\n",
       "                      ...  \n",
       "maize                 31335\n",
       "switchgrass           40001\n",
       "pig                   50247\n",
       "cattle                61859\n",
       "human                127314\n",
       "Length: 94, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_table('data/metadata.tsv.gz')\n",
    "metadata.rename({'sample_accession': 'sample'}, axis=1, inplace=True)\n",
    "nf = metadata[['sample', 'host_common_name']]\n",
    "nf = nf.merge(on='sample', right=data)\n",
    "nf[['host_common_name', 'amp']].drop_duplicates().groupby('host_common_name').agg('size').sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0e7c191d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "nenvo\n",
       "rabbit                    1\n",
       "guinea pig                1\n",
       "white-tailed deer         2\n",
       "fisher                    4\n",
       "wasp                      5\n",
       "                      ...  \n",
       "chicken               18259\n",
       "pig                   50247\n",
       "cattle                61859\n",
       "plant                103023\n",
       "human                127314\n",
       "Length: 70, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nf['nenvo'] = [x if x not in plants else 'plant' for x in nf.host_common_name]\n",
    "nf[['nenvo', 'amp']].drop_duplicates().groupby('nenvo').agg('size').sort_values()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "172a5fba",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = data[['sample', 'higher']]\n",
    "samples = samples.drop_duplicates()\n",
    "samples = samples.groupby('higher')\n",
    "samples = samples.agg('size')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c52c08b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "habitats = data[['higher', 'general_envo_name']].drop_duplicates()\n",
    "habitats = habitats.groupby('higher')['general_envo_name'].apply(lambda x: listing(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dc199ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "redamps = data.groupby('higher').agg('size')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d3a9f38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "nramps = data[['higher', 'amp']]\n",
    "nramps = nramps.drop_duplicates()\n",
    "nramps = nramps.groupby('higher')\n",
    "nramps = nramps.agg('size')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4bee8aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fams = pd.read_table('data/SPHERE_v.2022-03.levels_assessment.tsv.gz')\n",
    "fams = fams.rename({'AMP accession': 'amp',\n",
    "                    'SPHERE_fam level III': 'family'},\n",
    "                   axis=1)\n",
    "                   \n",
    "fams = fams[['amp', 'family']]\n",
    "data = data.merge(on='amp', right=fams)\n",
    "fams = fams.groupby('family').agg('size')\n",
    "fams = fams[fams >= 8].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cb5b436c",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data[['higher', 'family']].drop_duplicates()\n",
    "famps = data.groupby('higher').agg('size')\n",
    "famp_l = data[data.family.isin(fams)].groupby('higher').agg('size')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdcd0734",
   "metadata": {},
   "source": [
    "Here, it follows the supplementary table with info about the samples, number of redundant and non-redundant AMPs, as well as the number of clusters and families each high-level habitat affiliates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "697a60e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>high level environment</th>\n",
       "      <th>habitats</th>\n",
       "      <th>samples</th>\n",
       "      <th>redundant AMPs</th>\n",
       "      <th>non-redundant AMPs</th>\n",
       "      <th>AMP clusters</th>\n",
       "      <th>AMP families</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>anthropogenic</td>\n",
       "      <td>wastewater, built environment, activated sludg...</td>\n",
       "      <td>6129</td>\n",
       "      <td>304382</td>\n",
       "      <td>85988</td>\n",
       "      <td>63824</td>\n",
       "      <td>3913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aquatic</td>\n",
       "      <td>water associated, groundwater, river associate...</td>\n",
       "      <td>5961</td>\n",
       "      <td>762508</td>\n",
       "      <td>213861</td>\n",
       "      <td>147828</td>\n",
       "      <td>5116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>32640</td>\n",
       "      <td>2167411</td>\n",
       "      <td>111946</td>\n",
       "      <td>63641</td>\n",
       "      <td>3304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-human mammal gut</td>\n",
       "      <td>pig gut, dog gut, cattle gut, mouse gut, cat g...</td>\n",
       "      <td>3535</td>\n",
       "      <td>629929</td>\n",
       "      <td>104559</td>\n",
       "      <td>68364</td>\n",
       "      <td>3339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>other</td>\n",
       "      <td>air, mock community, algae associated, mine, i...</td>\n",
       "      <td>1300</td>\n",
       "      <td>39127</td>\n",
       "      <td>15609</td>\n",
       "      <td>13435</td>\n",
       "      <td>1306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>other animal</td>\n",
       "      <td>chicken gut, bird gut, cattle rumen, bee gut, ...</td>\n",
       "      <td>2356</td>\n",
       "      <td>348985</td>\n",
       "      <td>98615</td>\n",
       "      <td>73723</td>\n",
       "      <td>3275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>other human</td>\n",
       "      <td>human skin, human urogenital tract, human mout...</td>\n",
       "      <td>4052</td>\n",
       "      <td>96290</td>\n",
       "      <td>20137</td>\n",
       "      <td>13196</td>\n",
       "      <td>1112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>soil/plant</td>\n",
       "      <td>soil, plant associated</td>\n",
       "      <td>5420</td>\n",
       "      <td>1078992</td>\n",
       "      <td>315628</td>\n",
       "      <td>197565</td>\n",
       "      <td>5552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  high level environment                                           habitats  \\\n",
       "0          anthropogenic  wastewater, built environment, activated sludg...   \n",
       "1                aquatic  water associated, groundwater, river associate...   \n",
       "2              human gut                                          human gut   \n",
       "3   non-human mammal gut  pig gut, dog gut, cattle gut, mouse gut, cat g...   \n",
       "4                  other  air, mock community, algae associated, mine, i...   \n",
       "5           other animal  chicken gut, bird gut, cattle rumen, bee gut, ...   \n",
       "6            other human  human skin, human urogenital tract, human mout...   \n",
       "7             soil/plant                             soil, plant associated   \n",
       "\n",
       "   samples  redundant AMPs  non-redundant AMPs  AMP clusters  AMP families  \n",
       "0     6129          304382               85988         63824          3913  \n",
       "1     5961          762508              213861        147828          5116  \n",
       "2    32640         2167411              111946         63641          3304  \n",
       "3     3535          629929              104559         68364          3339  \n",
       "4     1300           39127               15609         13435          1306  \n",
       "5     2356          348985               98615         73723          3275  \n",
       "6     4052           96290               20137         13196          1112  \n",
       "7     5420         1078992              315628        197565          5552  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([habitats,\n",
    "                samples,\n",
    "                redamps,\n",
    "                nramps,\n",
    "                famps,\n",
    "                famp_l],\n",
    "               axis=1)\n",
    "\n",
    "df = df.reset_index()\n",
    "df = df.rename({'higher': 'high level environment',\n",
    "                'general_envo_name': 'habitats',\n",
    "                0: 'samples',\n",
    "                1: 'redundant AMPs',\n",
    "                2: 'non-redundant AMPs',\n",
    "                3: 'AMP clusters',\n",
    "                4: 'AMP families'},\n",
    "               axis=1)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac724b3d",
   "metadata": {},
   "source": [
    "### Information about the samples used in AMPSphere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bbc33b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data again\n",
    "data = pd.read_table('data/samples-min500k-assembly-prodigal-stats.tsv.gz')\n",
    "amps = pd.read_table('data/gmsc_amp_genes_envohr_source.tsv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "73781cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter columns\n",
    "namps = amps[['amp',\n",
    "              'sample',\n",
    "              'general_envo_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "283037bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# eliminate redundancy\n",
    "namps = namps.drop_duplicates()\n",
    "namps = namps.groupby('sample')\n",
    "namps = namps.agg('size')\n",
    "namps = namps.reset_index()\n",
    "\n",
    "namps = namps.rename({0: 'amps',\n",
    "                      'sample': 'sample_accession'},\n",
    "                     axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8f65d5a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge splitted data\n",
    "a = data.merge(on='sample_accession',\n",
    "               right=namps)\n",
    "               \n",
    "b = data[~data.sample_accession.isin(namps.sample_accession)]\n",
    "data = pd.concat([a, b])\n",
    "data.amps = data.amps.fillna(0)\n",
    "\n",
    "data['amps_per_assembled_Mbp'] = data.amps * 1_000_000 / data.assembly_total_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "52f98e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# more data...\n",
    "envo = pd.read_table('data/reduced_metadata.tsv.gz')\n",
    "data = data.merge(on='sample_accession', right=envo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ab5533d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>habitat</th>\n",
       "      <th>raw inserts</th>\n",
       "      <th>assembled bp</th>\n",
       "      <th>N50</th>\n",
       "      <th>ORFs+smORFs</th>\n",
       "      <th>smORFs</th>\n",
       "      <th>non-redundant AMPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Karasov_2018_arabidopsis_NextMet1</td>\n",
       "      <td>plant associated</td>\n",
       "      <td>12374844</td>\n",
       "      <td>6714448</td>\n",
       "      <td>700</td>\n",
       "      <td>10983</td>\n",
       "      <td>2254</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Karasov_2018_arabidopsis_NextMet124</td>\n",
       "      <td>plant associated</td>\n",
       "      <td>17119714</td>\n",
       "      <td>29141743</td>\n",
       "      <td>684</td>\n",
       "      <td>53733</td>\n",
       "      <td>10438</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Karasov_2018_arabidopsis_NextMet15</td>\n",
       "      <td>plant associated</td>\n",
       "      <td>13384712</td>\n",
       "      <td>3583114</td>\n",
       "      <td>1325</td>\n",
       "      <td>3660</td>\n",
       "      <td>1046</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Karasov_2018_arabidopsis_NextMet25</td>\n",
       "      <td>plant associated</td>\n",
       "      <td>15623952</td>\n",
       "      <td>5459772</td>\n",
       "      <td>736</td>\n",
       "      <td>7794</td>\n",
       "      <td>1998</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Karasov_2018_arabidopsis_NextMet50</td>\n",
       "      <td>plant associated</td>\n",
       "      <td>14905649</td>\n",
       "      <td>4297904</td>\n",
       "      <td>982</td>\n",
       "      <td>5138</td>\n",
       "      <td>1296</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63405</th>\n",
       "      <td>SAMEA6080433</td>\n",
       "      <td>human mouth</td>\n",
       "      <td>5013002</td>\n",
       "      <td>10785166</td>\n",
       "      <td>648</td>\n",
       "      <td>21077</td>\n",
       "      <td>6318</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63406</th>\n",
       "      <td>mgs635492</td>\n",
       "      <td>soil</td>\n",
       "      <td>25418882</td>\n",
       "      <td>6843991</td>\n",
       "      <td>369</td>\n",
       "      <td>17917</td>\n",
       "      <td>3077</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63407</th>\n",
       "      <td>mgs635439</td>\n",
       "      <td>soil</td>\n",
       "      <td>4516768</td>\n",
       "      <td>758472</td>\n",
       "      <td>273</td>\n",
       "      <td>2606</td>\n",
       "      <td>243</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63408</th>\n",
       "      <td>mgs635565</td>\n",
       "      <td>soil</td>\n",
       "      <td>1781932</td>\n",
       "      <td>41679</td>\n",
       "      <td>256</td>\n",
       "      <td>145</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63409</th>\n",
       "      <td>mgs635558</td>\n",
       "      <td>soil</td>\n",
       "      <td>15683066</td>\n",
       "      <td>3327349</td>\n",
       "      <td>273</td>\n",
       "      <td>11137</td>\n",
       "      <td>1218</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63410 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    sample           habitat  raw inserts  \\\n",
       "0        Karasov_2018_arabidopsis_NextMet1  plant associated     12374844   \n",
       "1      Karasov_2018_arabidopsis_NextMet124  plant associated     17119714   \n",
       "2       Karasov_2018_arabidopsis_NextMet15  plant associated     13384712   \n",
       "3       Karasov_2018_arabidopsis_NextMet25  plant associated     15623952   \n",
       "4       Karasov_2018_arabidopsis_NextMet50  plant associated     14905649   \n",
       "...                                    ...               ...          ...   \n",
       "63405                         SAMEA6080433       human mouth      5013002   \n",
       "63406                            mgs635492              soil     25418882   \n",
       "63407                            mgs635439              soil      4516768   \n",
       "63408                            mgs635565              soil      1781932   \n",
       "63409                            mgs635558              soil     15683066   \n",
       "\n",
       "       assembled bp   N50  ORFs+smORFs  smORFs  non-redundant AMPs  \n",
       "0           6714448   700        10983    2254                 2.0  \n",
       "1          29141743   684        53733   10438                 9.0  \n",
       "2           3583114  1325         3660    1046                 2.0  \n",
       "3           5459772   736         7794    1998                 5.0  \n",
       "4           4297904   982         5138    1296                 4.0  \n",
       "...             ...   ...          ...     ...                 ...  \n",
       "63405      10785166   648        21077    6318                 0.0  \n",
       "63406       6843991   369        17917    3077                 0.0  \n",
       "63407        758472   273         2606     243                 0.0  \n",
       "63408         41679   256          145       9                 0.0  \n",
       "63409       3327349   273        11137    1218                 0.0  \n",
       "\n",
       "[63410 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# supp table S1\n",
    "sup1 = data[['sample_accession', 'general_envo_name',\n",
    "             'inserts_raw', 'assembly_total_length',\n",
    "             'assembly_N50', 'prodigal_total_orfs',\n",
    "             'smORFs', 'amps']].copy()\n",
    "\n",
    "sup1.columns = ['sample', 'habitat', 'raw inserts',\n",
    "                'assembled bp', 'N50', 'ORFs+smORFs',\n",
    "                'smORFs', 'non-redundant AMPs']\n",
    "\n",
    "sup1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
