{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8568898d",
   "metadata": {},
   "source": [
    "# AMPSphere v.2022-03\n",
    "\n",
    "This is a notebook meant to form the set of notebooks used to analyze the data in AMPSphere and write the manuscript:\n",
    "\n",
    "__AMPSphere: Global survey of prokaryotic antimicrobial peptides shaping microbiomes__\n",
    "\n",
    "### Check the c_AMP density of species through different environments\n",
    "\n",
    "We generate sets of c_AMPs by species and sample. Then group them by environment and select those species happening in at least 2 habitats with at least 10 samples in each. The number of non-redundant c_AMPs attributed to a given species per sample in each sample is compared using Mann-Whitney U test and later the p-values are adjusted using Holm-Sidak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8fb71654",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from itertools import combinations, chain\n",
    "from scipy.stats import spearmanr, mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests as mt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5707b3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data = pd.read_table('../data_folder/complete_amps_associated_taxonomy.tsv.gz')\n",
    "metadata = pd.read_table(\"../data_folder/metadata.tsv.xz\")\n",
    "metadata = metadata.rename({'sample_accession': 'sample'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03b47dbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter species-specific genes\n",
    "data = data.loc[data.level == 'species', ['sample', 'source']]\n",
    "data = data.groupby(['sample', 'source']).agg('size')\n",
    "data = data.reset_index()\n",
    "data = data.rename({'source': 'name',\n",
    "                    0: 'amp_genes'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "173962e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genenvo = pd.read_table('../data_folder/general_envo_names.tsv.xz')\n",
    "\n",
    "metadata = metadata.merge(on=['microontology', 'host_scientific_name', 'host_tax_id'],\n",
    "                          right=genenvo,\n",
    "                          how='outer')\n",
    "\n",
    "metadata = metadata[['sample', 'general_envo_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee6d94fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate densities\n",
    "sps = pd.read_table('../data_folder/bps-per-sample-per-species.tsv.xz',\n",
    "                    sep='\\t',\n",
    "                    header='infer')\n",
    "\n",
    "data = data.merge(on=['sample', 'name'], right=sps)\n",
    "\n",
    "data['amp_density'] = data['amp_genes'] * 1e6 / data['nbps']\n",
    "\n",
    "# add metadata\n",
    "data = data.merge(on='sample', right=metadata)\n",
    "data.general_envo_name = data.general_envo_name.replace('human saliva', 'human mouth')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7e7b5a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tukeyfence(df):\n",
    "    itdf = pd.DataFrame()\n",
    "    for record in df.groupby(['name', 'general_envo_name']):\n",
    "        record = record[1]\n",
    "        if len(record) > 9:\n",
    "            q1, q3 = record['amp_density'].quantile([0.25,0.75])\n",
    "            iqr = q3-q1\n",
    "            ul, ll = q3+(1.5*iqr), q1-(1.5*iqr)\n",
    "            record = record[(record.amp_density <= ul) & (record.amp_density >= ll)]\n",
    "            itdf = pd.concat([itdf, record])\n",
    "    return itdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7663382d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>name</th>\n",
       "      <th>amp_genes</th>\n",
       "      <th>taxid</th>\n",
       "      <th>nbps</th>\n",
       "      <th>amp_density</th>\n",
       "      <th>general_envo_name</th>\n",
       "      <th>moe</th>\n",
       "      <th>moe_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>901690</th>\n",
       "      <td>SAMN06264886</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>2</td>\n",
       "      <td>44747</td>\n",
       "      <td>1128853</td>\n",
       "      <td>1.771710</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.455462</td>\n",
       "      <td>138.592806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>924151</th>\n",
       "      <td>SAMN06266445</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>1</td>\n",
       "      <td>44747</td>\n",
       "      <td>767610</td>\n",
       "      <td>1.302745</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.553378</td>\n",
       "      <td>195.999872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989431</th>\n",
       "      <td>SAMN06344006</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>1</td>\n",
       "      <td>44747</td>\n",
       "      <td>613652</td>\n",
       "      <td>1.629588</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>3.193990</td>\n",
       "      <td>195.999840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1158011</th>\n",
       "      <td>SAMN07960327</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>2</td>\n",
       "      <td>44747</td>\n",
       "      <td>1089727</td>\n",
       "      <td>1.835322</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.543624</td>\n",
       "      <td>138.592802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1158180</th>\n",
       "      <td>SAMN07960328</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>2</td>\n",
       "      <td>44747</td>\n",
       "      <td>1100407</td>\n",
       "      <td>1.817509</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.518937</td>\n",
       "      <td>138.592803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1660322</th>\n",
       "      <td>TARA_142_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>4485880</td>\n",
       "      <td>0.222922</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.436926</td>\n",
       "      <td>195.999978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1660513</th>\n",
       "      <td>TARA_145_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>4258030</td>\n",
       "      <td>0.234850</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.460307</td>\n",
       "      <td>195.999977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1660989</th>\n",
       "      <td>TARA_148b_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>3</td>\n",
       "      <td>1762</td>\n",
       "      <td>3802238</td>\n",
       "      <td>0.789009</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.892847</td>\n",
       "      <td>113.160608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1661071</th>\n",
       "      <td>TARA_149_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>3801975</td>\n",
       "      <td>0.263021</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.515521</td>\n",
       "      <td>195.999974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1661703</th>\n",
       "      <td>TARA_152_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>3417823</td>\n",
       "      <td>0.292584</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.573464</td>\n",
       "      <td>195.999971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1365285 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sample                         name  amp_genes  \\\n",
       "901690             SAMN06264886  0-14-0-20-30-16 sp002779075          2   \n",
       "924151             SAMN06266445  0-14-0-20-30-16 sp002779075          1   \n",
       "989431             SAMN06344006  0-14-0-20-30-16 sp002779075          1   \n",
       "1158011            SAMN07960327  0-14-0-20-30-16 sp002779075          2   \n",
       "1158180            SAMN07960328  0-14-0-20-30-16 sp002779075          2   \n",
       "...                         ...                          ...        ...   \n",
       "1660322   TARA_142_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "1660513   TARA_145_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "1660989  TARA_148b_MES_0.22-3_G        Zunongwangia profunda          3   \n",
       "1661071   TARA_149_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "1661703   TARA_152_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "\n",
       "         taxid     nbps  amp_density general_envo_name       moe     moe_pct  \n",
       "901690   44747  1128853     1.771710       groundwater  2.455462  138.592806  \n",
       "924151   44747   767610     1.302745       groundwater  2.553378  195.999872  \n",
       "989431   44747   613652     1.629588       groundwater  3.193990  195.999840  \n",
       "1158011  44747  1089727     1.835322       groundwater  2.543624  138.592802  \n",
       "1158180  44747  1100407     1.817509       groundwater  2.518937  138.592803  \n",
       "...        ...      ...          ...               ...       ...         ...  \n",
       "1660322   1762  4485880     0.222922            marine  0.436926  195.999978  \n",
       "1660513   1762  4258030     0.234850            marine  0.460307  195.999977  \n",
       "1660989   1762  3802238     0.789009            marine  0.892847  113.160608  \n",
       "1661071   1762  3801975     0.263021            marine  0.515521  195.999974  \n",
       "1661703   1762  3417823     0.292584            marine  0.573464  195.999971  \n",
       "\n",
       "[1365285 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove outliers\n",
    "df = tukeyfence(data)\n",
    "df['p'] = df.amp_density / 1e6\n",
    "df['moe'] = np.sqrt(df.p*(1-df.p)/df.nbps) * 1.96 * 1e6\n",
    "df.drop('p', axis=1, inplace=True)\n",
    "df['moe_pct'] = df.moe * 100 / df.amp_density\n",
    "df.to_csv('species_amp_density_per_sample.tsv.gz',\n",
    "          sep='\\t',\n",
    "          header=True,\n",
    "          index=None)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1653ed08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select species present in >=10 samples per habitats\n",
    "k = df.groupby(['name', 'general_envo_name']).agg('size')\n",
    "k = k[k >= 10]\n",
    "\n",
    "# select species present in more than 1 habitat\n",
    "k = k.reset_index().groupby('name').apply(lambda x: x.general_envo_name.tolist())\n",
    "k = k[k.apply(lambda x: len(x)) > 1]\n",
    "k = k.reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30ccefdf",
   "metadata": {},
   "source": [
    "### Supplementary Table S8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2f719233",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/celio/miniconda3/lib/python3.9/site-packages/statsmodels/stats/multitest.py:177: RuntimeWarning: divide by zero encountered in log1p\n",
      "  np.log1p(-pvals))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>env1</th>\n",
       "      <th>env1_samples</th>\n",
       "      <th>amp_density_avg_e1</th>\n",
       "      <th>amp_density_std_e1</th>\n",
       "      <th>env2</th>\n",
       "      <th>env2_samples</th>\n",
       "      <th>amp_density_avg_e2</th>\n",
       "      <th>amp_density_std_e2</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>27</td>\n",
       "      <td>1.065536</td>\n",
       "      <td>0.559337</td>\n",
       "      <td>marine</td>\n",
       "      <td>25</td>\n",
       "      <td>0.987186</td>\n",
       "      <td>0.569810</td>\n",
       "      <td>6.209545e-01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>27</td>\n",
       "      <td>1.065536</td>\n",
       "      <td>0.559337</td>\n",
       "      <td>soil</td>\n",
       "      <td>155</td>\n",
       "      <td>1.207563</td>\n",
       "      <td>0.834207</td>\n",
       "      <td>7.908466e-01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>27</td>\n",
       "      <td>1.065536</td>\n",
       "      <td>0.559337</td>\n",
       "      <td>water associated</td>\n",
       "      <td>45</td>\n",
       "      <td>0.810040</td>\n",
       "      <td>0.405034</td>\n",
       "      <td>6.273530e-02</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>marine</td>\n",
       "      <td>25</td>\n",
       "      <td>0.987186</td>\n",
       "      <td>0.569810</td>\n",
       "      <td>soil</td>\n",
       "      <td>155</td>\n",
       "      <td>1.207563</td>\n",
       "      <td>0.834207</td>\n",
       "      <td>4.466062e-01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>marine</td>\n",
       "      <td>25</td>\n",
       "      <td>0.987186</td>\n",
       "      <td>0.569810</td>\n",
       "      <td>water associated</td>\n",
       "      <td>45</td>\n",
       "      <td>0.810040</td>\n",
       "      <td>0.405034</td>\n",
       "      <td>2.296780e-01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9927</th>\n",
       "      <td>Zag111 sp004558955</td>\n",
       "      <td>human gut</td>\n",
       "      <td>55</td>\n",
       "      <td>161.060656</td>\n",
       "      <td>90.532428</td>\n",
       "      <td>primate gut</td>\n",
       "      <td>15</td>\n",
       "      <td>21.175229</td>\n",
       "      <td>18.581715</td>\n",
       "      <td>8.012767e-09</td>\n",
       "      <td>0.000068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9928</th>\n",
       "      <td>Zag111 sp004558955</td>\n",
       "      <td>pig gut</td>\n",
       "      <td>145</td>\n",
       "      <td>3.867173</td>\n",
       "      <td>3.154102</td>\n",
       "      <td>primate gut</td>\n",
       "      <td>15</td>\n",
       "      <td>21.175229</td>\n",
       "      <td>18.581715</td>\n",
       "      <td>6.541202e-05</td>\n",
       "      <td>0.335970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9929</th>\n",
       "      <td>Zavarzinella formosa</td>\n",
       "      <td>plant associated</td>\n",
       "      <td>15</td>\n",
       "      <td>1.579412</td>\n",
       "      <td>1.080763</td>\n",
       "      <td>soil</td>\n",
       "      <td>123</td>\n",
       "      <td>1.602291</td>\n",
       "      <td>1.331643</td>\n",
       "      <td>7.843780e-01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9930</th>\n",
       "      <td>Zoogloea ramigera</td>\n",
       "      <td>activated sludge</td>\n",
       "      <td>113</td>\n",
       "      <td>1.057880</td>\n",
       "      <td>0.688160</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>167</td>\n",
       "      <td>1.132519</td>\n",
       "      <td>0.752228</td>\n",
       "      <td>4.592241e-01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9931</th>\n",
       "      <td>Zoogloea sp002028455</td>\n",
       "      <td>activated sludge</td>\n",
       "      <td>68</td>\n",
       "      <td>1.338772</td>\n",
       "      <td>0.683347</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>93</td>\n",
       "      <td>2.512038</td>\n",
       "      <td>2.536020</td>\n",
       "      <td>1.879802e-02</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9932 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          species              env1  env1_samples  \\\n",
       "0     0-14-0-80-60-11 sp002779455       groundwater            27   \n",
       "1     0-14-0-80-60-11 sp002779455       groundwater            27   \n",
       "2     0-14-0-80-60-11 sp002779455       groundwater            27   \n",
       "3     0-14-0-80-60-11 sp002779455            marine            25   \n",
       "4     0-14-0-80-60-11 sp002779455            marine            25   \n",
       "...                           ...               ...           ...   \n",
       "9927           Zag111 sp004558955         human gut            55   \n",
       "9928           Zag111 sp004558955           pig gut           145   \n",
       "9929         Zavarzinella formosa  plant associated            15   \n",
       "9930            Zoogloea ramigera  activated sludge           113   \n",
       "9931         Zoogloea sp002028455  activated sludge            68   \n",
       "\n",
       "      amp_density_avg_e1  amp_density_std_e1              env2  env2_samples  \\\n",
       "0               1.065536            0.559337            marine            25   \n",
       "1               1.065536            0.559337              soil           155   \n",
       "2               1.065536            0.559337  water associated            45   \n",
       "3               0.987186            0.569810              soil           155   \n",
       "4               0.987186            0.569810  water associated            45   \n",
       "...                  ...                 ...               ...           ...   \n",
       "9927          161.060656           90.532428       primate gut            15   \n",
       "9928            3.867173            3.154102       primate gut            15   \n",
       "9929            1.579412            1.080763              soil           123   \n",
       "9930            1.057880            0.688160        wastewater           167   \n",
       "9931            1.338772            0.683347        wastewater            93   \n",
       "\n",
       "      amp_density_avg_e2  amp_density_std_e2       p_value     p_adj  \n",
       "0               0.987186            0.569810  6.209545e-01  1.000000  \n",
       "1               1.207563            0.834207  7.908466e-01  1.000000  \n",
       "2               0.810040            0.405034  6.273530e-02  1.000000  \n",
       "3               1.207563            0.834207  4.466062e-01  1.000000  \n",
       "4               0.810040            0.405034  2.296780e-01  1.000000  \n",
       "...                  ...                 ...           ...       ...  \n",
       "9927           21.175229           18.581715  8.012767e-09  0.000068  \n",
       "9928           21.175229           18.581715  6.541202e-05  0.335970  \n",
       "9929            1.602291            1.331643  7.843780e-01  1.000000  \n",
       "9930            1.132519            0.752228  4.592241e-01  1.000000  \n",
       "9931            2.512038            2.536020  1.879802e-02  1.000000  \n",
       "\n",
       "[9932 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test difference in the amp_densitys among samples of\n",
    "# the same species through different environments\n",
    "test = []\n",
    "for _, s, env in k.itertuples():\n",
    "    combs = combinations(env, 2)\n",
    "    for i, j in combs:\n",
    "        n1 = df[(df.name == s) & (df.general_envo_name == i)]\n",
    "        n2 = df[(df.name == s) & (df.general_envo_name == j)]\n",
    "        u, p = mannwhitneyu(n1['amp_density'], n2['amp_density'])\n",
    "        test.append((s, i, len(n1),\n",
    "                     n1.amp_density.mean(), n1.amp_density.std(),\n",
    "                     j, len(n2), n2.amp_density.mean(),\n",
    "                     n2.amp_density.std(), p))\n",
    "\n",
    "# format results\n",
    "test = pd.DataFrame(test, columns=['species',\n",
    "                                   'env1',\n",
    "                                   'env1_samples',\n",
    "                                   'amp_density_avg_e1',\n",
    "                                   'amp_density_std_e1',\n",
    "                                   'env2',\n",
    "                                   'env2_samples',\n",
    "                                   'amp_density_avg_e2',\n",
    "                                   'amp_density_std_e2',\n",
    "                                   'p_value'])\n",
    "\n",
    "test\n",
    "\n",
    "# adjust p-values\n",
    "_, test['p_adj'], _, _ = mt(test['p_value'], method='hs')\n",
    "\n",
    "# export\n",
    "test.to_csv('species_amp_density_crossenvironment.tsv.gz', sep='\\t', header=True, index=None)\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7e5281a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It was tested 3930 species\n",
      "Only 38.96% (1531) of them\n",
      "presented significant differences in their AMP density\n",
      "across different environments\n",
      "In average, it was observed 2.3e+01 c_AMP genes per assembled Mbp for each species\n",
      "per sample, ranging between 2.329e+03-0.13.\n"
     ]
    }
   ],
   "source": [
    "### Generating statistics\n",
    "\n",
    "# select environments with at least 100 samples containing AMPs\n",
    "k = data[['sample', 'general_envo_name']].drop_duplicates()\n",
    "k = k.general_envo_name.value_counts()\n",
    "k = k[k>=100].index\n",
    "\n",
    "# select pairs containing both environments in the list\n",
    "x = test[(test.env1.isin(k)) & (test.env2.isin(k))]\n",
    "\n",
    "n1 = len(set(x.species))\n",
    "print(f'It was tested {n1} species')\n",
    "\n",
    "n2 = len(set(x.loc[test.p_adj < 0.05, 'species']))\n",
    "print(f'''Only {round((n2*100/n1), 2)}% ({n2}) of them\n",
    "presented significant differences in their AMP density\n",
    "across different environments''')\n",
    "\n",
    "savg = (x.amp_density_avg_e1.sum() + x.amp_density_avg_e2.sum()) / (2*len(x))\n",
    "\n",
    "mmax, mmin = [max(x.amp_density_avg_e1.max(), x.amp_density_avg_e2.max()),\n",
    "              min(x.amp_density_avg_e1.min(), x.amp_density_avg_e2.min())]\n",
    "\n",
    "print(f'''In average, it was observed {savg:.2} c_AMP genes per assembled Mbp for each species\n",
    "per sample, ranging between {mmax:.4}-{mmin:.2}.''')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
